python/tank/platform/engine.py

```diff
@@ -565,63 +566,52 @@ def execute_in_main_thread(self, func, *args, **kwargs):
-            return v
-
-        def threaded_work(val):
-            eng = sgtk.platform.current_engine()
-            c_time = 0.0
-            for i in range(NUM_THREAD_ITERATIONS):
-                time.sleep(random.randint(0, 10)/10.0)
-                arg = (val, i)
-                st = time.time()
-                ret_val = eng.execute_in_main_thread(run_in_main_thread, arg)
-                e = time.time()
-                #print "Time to run: %0.4fs" % (e-st)
-                c_time += (e-st)
-                if ret_val != arg:
-                    print "Result mismatch: %s != %s!!" % (ret_val, arg)
-
-            print "Cumulative time for thread %d: %0.4fs" % (val, c_time)
-
-        threads = []
-        for ti in range(NUM_TEST_THREADS):
-            t = threading.Thread(target=lambda:threaded_work(ti))
-            t.start()
-            threads.append(t)
-        # ---------------------------------------------------------------------------------
-        # ---------------------------------------------------------------------------------
-
         :param func: function to call
         :param args: arguments to pass to the function
         :param kwargs: named arguments to pass to the function
 
         :returns: the result of the function call
         """
-        if self._invoker:
+        return self._execute_in_main_thread("_invoker", func, *args, **kwargs)
+
+    def async_execute_in_main_thread(self, func, *args, **kwargs):
+        """
+        Execute the specified function in the main thread when called from a non-main
+        thread.  This call will return immediately and will not wait for the code to be
+        executed in the main thread.
+
+        Note, this currently only works if Qt is available, otherwise it just
+        executes on the current thread.
+
+        :param func: function to call
+        :param args: arguments to pass to the function
+        :param kwargs: named arguments to pass to the function
+        """
+        self._execute_in_main_thread("_async_invoker", func, *args, **kwargs)
+
+    def _execute_in_main_thread(self, invoker_name, func, *args, **kwargs):
+        """
+        Executes the given method and arguments with the specified invoker.
+        If the invoker is not ready or if the calling thread is the main thread,
+        the method is called immediately with it's arguments.
+
+        :param invoker_name: Name of the invoker to use.
+        :param func: function to call
+        :param args: arguments to pass to the function
+        :param kwargs: named arguments to pass to the function
+
+        :returns: The return value from the invoker.
+        """
+        # Execute in main thread might be called before the invoker is ready.
+        # For example, an engine might use the invoker for logging to the main
+        # thread.
+        invoker = getattr(self, invoker_name) if hasattr(self, invoker_name) else None
```

Comment:
following up from the above comment, if we had the enum approach above it means we don't have to do this `getattr`-type introspection.